src/cufinufft.cu

```diff
@@ -1,17 +1,74 @@
 #include <iostream>
 #include <iomanip>
+#include <vector>
+#include <complex>
+#include <cstddef>
 #include <math.h>
 #include <helper_cuda.h>
 #include <complex>
 #include <cufft.h>
+// #include <thrust/zip_iterator.h>
+#include <thrust/tuple.h>
+#include <thrust/transform.h>
+#include <thrust/execution_policy.h>
+#include <thrust/iterator/zip_iterator.h>
+#include <thrust/tuple.h>
+#include <thrust/complex.h>
 
 #include <cufinufft_eitherprec.h>
 #include "cuspreadinterp.h"
 #include "cudeconvolve.h"
 #include "memtransfer.h"
+#include "util_t3.h"
 
 using namespace std;
 
+template <typename T>
```

Comment:
I'm not familiar with this construction. Can you explain briefly what's going on? Why is it necessary for the type 3 code?

Reply:
These structs are used for applying scaling and prephase required for type 3. To avoid having to write kernel launch logic by hand, I'm using `thrust::transform` together with these structs. By using templates, no additional macro naming magic is required to prevent symbol collision between single and double precision object files.

Reply:
Ok. It would be good to keep these things consistent across the code, however. We have to think about maintenance down the road. If this is only to take care of single vs. double, would you mind using the macro system?

Reply:
Decided to go with templating here. Closing.